server/static/submit.jsx

```diff
@@ -0,0 +1,155 @@
+        this.handleSubmit = this.handleSubmit.bind(this);
+    }
+
+    resetState(error) {
+        this.setState({ git_url: '', git_user: '', project_name: '', valid: false, error: error });
+    }
+
+    handleChange(event) {
+        var re = RegExp('^(?:https:\/\/|http:\/\/|)(?:www.|)github.com\/([a-z0-9]*)\/([a-z0-9-_]*)(?:\/|)(?:.git|)$')
+
+        var git_user
+        var project_name
+
+        var github_url_candidate = event.target.value
+        if (re.test(github_url_candidate)) {
+            var groups = github_url_candidate.match(re)
+
+            git_user = groups[1]
+            project_name = groups[2]
+
+            this.setState({
+                git_url: groups[0],
+                git_user: git_user,
+                project_name: project_name,
+                error: "none",
+                valid: true
+            });
+        } else {
+            this.resetState("invalid_url")
+            return
+        }
+        /*
+        fetch(`https://api.github.com/repos/${git_user}/${project_name}/contents`)
+            .then(
+                (result) => {
+                    console.log(result)
+                    if (result.ok) {
+                        result.json().then((data) => {
+                            if (!this.run_exists(data)) {
+                                this.setState({
+                                    valid: false,
+                                    error: "no_run"
+                                });
+                            } else {
+                                this.setState({
+                                    valid: true,
+                                    error: "none"
+                                });
+                            }
+                        })
+                    } else {
+                        this.setState({
+                            valid: false,
+                            error: "repo_not_found"
+                        });
+                    }
+                }
+            ).catch(
+                (error) => {
+                    this.setState({
+                        valid: false,
+                        error: "unknown"
+                    });
+                }
+            )
+            */
+
+    }
+
+    run_exists(files) {
+        // returns if the file list contains run.py
+        var found = false;
+        for (var i = 0, len = files.length; i < len; i++) {
+            if (files[i].path == "run.py") {
+                found = true;
+            }
+        }
+        return found
+    }
+
+    handleSubmit(event) {
+        event.preventDefault();
+        var form_data = new FormData();
+
+        for (var key in this.state) {
+            form_data.append(key, this.state[key]);
+        }
+
+        fetch('/job', {
+            method: 'POST',
+            body: form_data,
+        });
+        window.document.location = "/"
+    }
+
+    render() {
+        return (
```

Comment:
All this render return jazz, can it be held somewhere else like we did with the html templates? Or is that dumb? Just thinking that this is very formatty of our content type code where the rest is queries and backend shit. Just to try and divide out backend vs frontend. Just a thought, not a real strong suggestion.